### DISCUSSION TITLE
Expanded MIDI implementation

### DISCUSSION THREAD

**Discussion by Smi23y** [bangcorrupt/hacktribe#287] [Q&A] [answered]
Hello, thanks for the great project.  I've got one more question.  Is it possible to extend the Midi implementation to include the “chain mode (on/off)” setting?

**Comment by bangcorrupt** [comment, marked as answer]
In the future I'd like to add MIDI control of everything.  For now we can change this using sysex read-modify-write.

**Reply by othmar52** [reply]
> ...For now we can change this using sysex read-modify-write.  

Are there any examples in the documentation/wiki how to achieve this?  
Unfortunately i didn't find anything  (the closest i found for sending/receiving SysEx is [hacktribe-editor](https://github.com/bangcorrupt/hacktribe-editor) and [e2sysex.py](https://github.com/bangcorrupt/hacktribe/wiki/Scripts#e2sysexpy) )  

Thanks in advance for pointing me in the right direction!

**Reply by bangcorrupt** [reply]
Only the [MIDI Implementation](https://cdn.korg.com/us/support/download/files/1d5b06f3afe80bb12274be4b6fca5dc8.txt?response-content-disposition=attachment%3Bfilename%2A%3DUTF-8%27%27electribe_sampler_MIDIimp.txt&response-content-type=application%2Foctet-stream%3B).  There is a breakdown of the file structure at the end of the document.  [ht_syx_codec.py](https://github.com/bangcorrupt/hacktribe-editor/blob/ht-cli/hacktribe_editor/ht_syx_codec.py) can convert between sysex and raw binary formats.

There is an option to enable chain mode in the global settings, or individually in each pattern file.  [e2sysex.py](https://github.com/bangcorrupt/hacktribe-editor/blob/main/e2sysex.py) on the `main` branch of Hacktribe Editor can read and write patterns, writing global settings isn't supported.  [ht_sysex.py](https://github.com/bangcorrupt/hacktribe-editor/blob/ht-cli/hacktribe_editor/ht_sysex.py) on the `ht-cli` branch of Hacktribe Editor will return sysex messages as bytes.  This may be easier to integrate with other tools.

There are a few other editor projects, I'm not sure of their features.

https://github.com/rafamj/elecmidi
https://github.com/konsumer/electribe_tools
https://github.com/apiel/electribe-ui
https://github.com/egasimus/dawless

**Reply by othmar52** [reply]
thx for your reply. i will give it a try

**Comment by JinKei91** [comment]
@bangcorrupt  Regarding midi implementation. On electribe2 every of 16 pads is static midi channel which we can not change, pad 4 is midi channel 4, pad 9 is midi channel 9 and so on. Is it possible to improve this? to do as modern devices can, so we can assign any midi channel to any of 16 pads, and for example on pad 1 we can assign midi channel 5 , on pad 9 midi channel 1, on pad 10 midi channel 1 as well and so on. So make midi implementation more flexible.

**Reply by bangcorrupt** [reply]
I'll try and make this configurable via NRPN, along with the CC numbers for each control.

**Reply by JinKei91** [reply]
any chance to try make midi channel configurable?
